### ISSUE TITLE
IO error

### ISSUE THREAD

**Issue by trollius** [GoogleCloudPlatform/gcsfuse#191]
Hello Aaron,

we have a problem with input/output error in gcsfuse with some files and this error appears frequently. Here is an exact trackback of what happens:
1. We have a bunch of files (1...20), among them is a file1 ([description of file1](https://gist.github.com/trollius/1779b30c9732285f1e716c7359087e79)).
2. The sizes of these files are 140, and they are very similar in their content
3. 14 minutes after creation of these files, there was a lockfile created ([description of lockfile](https://gist.github.com/trollius/8ee9f72b3eac9b02de3f44ace660d790))
4. The a backup file for file1 is created and a tool which supposed to use this file does a backup query to this file ([logfile1 is here](https://gist.github.com/trollius/78ee50780f6e2cb404434f24da32c0c0))
5. Google returns 200 and content-length is 140 (as it should be). We also [get following file](http://prnt.sc/ckzjhy)
6. And finally right after that we get 404 error and IO error ([logfile2 is here](https://gist.github.com/trollius/eead684a20c7e6351e41cbfdfa44e311))

So in summary: small file 140bytes is created, then lockfile is created, then there is a query to a backup file (which returns 200 status - OK) and then there is a second query to this backup file (which returns status 404 - not OK) and the entire operation fails.

Do you have any ideas what's causing it? we usually have 2-3 such cases in 20 test runs.

We have version  gcsfuse 15.1, because newer releases have about 7-8 such cases in 20 test runs

Best,
Anna

**Comment by jacobsa** [comment]
Hi, a few things:

1. Please update to the latest version for the purposes of this discussion; trying to reason about old code makes this much more difficult.

2. We previously discussed lock files in #178, where I concluded they probably weren't going to work correctly due to gcsfuse not supporting `link(2)`. Is this not the case? Please be careful about relying on gcsfuse for lock files unless you're quite sure you understand the risks.

3. I'll proceed assuming you've mixed up your log files -- the first shows the 404 and the second doesn't.

4. As far as I can tell, `logfile2` shows a 404 for `RNG_state_0027.bin`, but neither log file shows that object being created. Are you sure it was created?

**Comment by trollius** [comment]
Hello Aaron,

1. yes, we will upgrade to a newer version
2. in our case lockfiles are simply `touch` for existing files (instead of lock)
3. correct
4. `RNG_state_0027.bin` was created before running the code so that's why creation is absent in `logfile2`

Do you need us to run the same processing with the newest gcsfuse version or there is enough of details for debagging already?

Anna

**Comment by jacobsa** [comment]
Upgrading versions would be a good idea just in case.

But I'm not sure what to say here except to check -- are you quite sure that an object named *exactly* `notcahelknatpasspr%2FRNG_state_0027.bin` already existed in the bucket? The [HTTP request sequence](https://gist.github.com/trollius/78ee50780f6e2cb404434f24da32c0c0#file-logfile2-L8-L28) in `logfile2` shows clearly that GCS thinks it doesn't exist.

**Comment by trollius** [comment]
hello Aaron,
In logfile1 you can see a request to a backup file and we got 200 response. Then there is a request to exactly same file and we get 404 error and weird symbols as in the picture. So yes, the file does seem to exist

**Comment by trollius** [comment]
Is there any additional logging we can try to do to catch at least something about this problem?

**Comment by jacobsa** [comment]
These are not exactly the same file:

1. In logfile1 the object is named `notcahelknatpasspr/RNG_state_0027.bin_backup`.
2. In logfile2 the object is named `notcahelknatpasspr/RNG_state_0027.bin`.

Are you sure there is an object named `notcahelknatpasspr/RNG_state_0027.bin`?

**Comment by oajamiehe** [comment]
i'm facing the same error intermittently when i tried for example to list the contents of the partition that is mounted on the bucket using gcfuse and to solve this i had to remount the partition again

**Comment by jacobsa** [comment]
@oajamiehe: could you please use `--foreground` and `--debug_fuse` to collect some logging?

**Comment by jacobsa** [comment]
@oajamiehe @trollius: are you still seeing issues? Could you please check that the object actually exists, and provide logging that proves it? It seems like perhaps there was just a confusion about the object name.

**Comment by trollius** [comment]
Hello Aaron,

We keep tracking the debug info but lately there were literally nothing to
catch. As soon as we catch something we will report immideately

Anna
